File: LC/[2158] Amount of New Area Painted Each Day.py
from collections import defaultdict
from heapq import heappush, heappop
from typing import List, DefaultDict, Tuple

class Solution:
    def amountPainted(self, paint: List[List[int]]) -> List[int]:
        N=len(paint)
        ## declare/build events [loc, [[dd, flag]]]
        events: DefaultDict[List[Tuple[int, int]]] = defaultdict(list)
        for dd, (st, ed) in enumerate(paint):
            events[st].append((dd, 1))
            events[ed].append((dd, -1))
        
        ret=[0]*N
        mnh: List[int]=[]
        day2freq=defaultdict(int)
        # for each location, find the effective paint day and update contribution
        sorted_loc = sorted(events.keys()) # sorted location
        for ii, loc in enumerate(sorted_loc):
            for [dd, flag] in events[loc]:
                if flag > 0:
                    heappush(mnh, dd)
                    day2freq[dd] += 1
                else:
                    day2freq[dd] -= 1
            
            while mnh and day2freq[mnh[0]] == 0:
                heappop(mnh)
            
            if mnh and ii+1 < len(sorted_loc):
                 ret[mnh[0]] += sorted_loc[ii+1] - loc

        return ret

File: LC/[2158] Amount of New Area Painted Each Day.cpp
class Solution {
public:
    vector<int> amountPainted(vector<vector<int>>& paint) {

        //given a position, collect all events with event day: {eventDay, event}
        map<int, vector<pair<int, int>>> Map; //pos->{{idx, flag}}
        //given an location, which days paints it?
        for(int ii=0; ii<paint.size(); ++ii){//Map collects all events given an position/location
            Map[paint[ii][0]].push_back({ii, 1});
            Map[paint[ii][1]].push_back({ii, -1});
        }


        //array: all sorted locations with paint event on some day 
        //events array: expands events sorted by position, position/location, each position, we have all events with date info
        vector<pair<int, vector<pair<int, int>>>> array(Map.begin(), Map.end());//cpp, expand map key/val to a vector
        set<int> Set;//to collect all paint events on some day and sorted by day. the min day is the effective paint
        vector<int> ret(paint.size());//return space (amount of new area painted on a day ii)

        for(int ii=0; ii<array.size(); ++ii){//for each position/location with event happened
            int pos = array[ii].first; //the position of this event
            for(auto& [idx, flag]: array[ii].second){//all event day/event at this position/location
                if(flag==1)
                    Set.insert(idx);//this set is sorted by day, the first date is the effective paint day
                else
                    Set.erase(idx);
            }
            if(!Set.empty())//Set not empty meaning the location is painted by multiple days, and the first day is the effective paint
                //until the next event, all painting belong to this day. we will evalute on next event point see which day contributed the painting
                ret[*Set.begin()] += array[ii+1].first - array[ii].first;//array is loc, (day, flag)
        }
        return ret;
    }
};